### DISCUSSION TITLE
How to understand the status code

### DISCUSSION THREAD

**Discussion by cigarl** [apache/iotdb#3844] [Ideas]
Hi,all

When I tried to fix bugs, I found that i am confused about the rules of the status code.  For the same problem, the status code returned to the client may be different, and there is no clear hierarchy.

For example, When _WAL_ failed, we will return _WRITE_PROCESS_ERROR_  in `insertRow`, and _INTERNAL_SERVER_ERROR_ in `insertTablet`,and i can't find the hierarchy between them. 

Just like the directory structure of code, we need  a clear hierarchy of response to client. Details of response in `BACTH_INSERT `should be in a same rank with the result of `INSERT`.

IMO, we need three levels at least, _ATOMIC_,_BATCH_ and _SERVER_. The status of _ATOMIC_ should belong to _BATCH_.And status codes of _SERVER_ should be carefully returned to the client,that means there may be a serious problem in server,just like _INTERNAL_SERVER_ERROR_ and _READ_ONLY_SYSTEM_ERROR_ etc.

Of course, we might also need the cluster level. 

Maybe, it needs everyone's opinions and ideas  : )

**Comment by jixuan1989** [comment]
Hi, 

IMO, 
if a write (no matter what operation is, either insertRow or insertTablet) fails 
because of users' invalid data, we need to return error codes like  401 (SQL parser error), 601 (no permission), 413 (write reject).

If the server has something wrong in the writing process,  412 (write process error) is returned.
If the server has something wrong that we do not know, 500 (server error) is returned.
